apps/dashr-studybrowser/app.R

```diff
@@ -0,0 +1,286 @@
+default_options = foreach(i=default_test_articles, combine='c') %do% returnOptions(i, default_study_data)
+default_options =  default_options[[1]]
+app <- Dash$new()
+
+default_plot <- plot_ly(default_study_input, type='violin', y= ~reading_value, name=~group_name, text=~subject_id, hoveron='points',  meanline=l.("visible"= T), showlegend=F,points='all', pointpos=0, color=~group_colors)
+
+app$layout(
+  
+  htmlDiv(className='', children=l.(
+    
+    #Title
+    htmlDiv(className='study-browser-banner', children="Animal Study Browser")
+    ,
+    htmlDiv(className='container', children= l.(
+      
+      htmlDiv(className='row', children= l.(
+        
+        htmlDiv(className='six columns', children=l.(
+          
+          htmlEm('Test Article:')
+          ,
+          dccDropdown(id='study-dropdown', options=l.(), value='')
+        
+        ))
+        ,
+        htmlDiv(className='six columns', children = l.())
+        
+        
+      ))
+      ,
+      htmlDiv(className='row', children=l.(
+        
+        htmlDiv(className='ten columns', children=l.(
+          
+          dccGraph(id='plot', figure=default_plot)
+          
+        )
+        
+       
+        )
+        ,
+        htmlDiv(className='two columns', children=l.(
+          
+          dccRadioItems(id='chart-type', options=l.(l.('label'='Box Plot', 'value'='box'), l.('label'='Violin Plot', 'value'='violin')),  value='violin')
+          
+        ))
+        
+      ))
+      ,
+      
+      htmlDiv(className='row', style=l.(), children=l.(
+        
+        htmlDiv(className='twelve columns', children=l.(
+          
+          dccUpload(
+            
+            id='upload-data'
+            ,
+            children= htmlDiv(l.("Drag and Drop or ", htmlA("Select Files")))
+            ,
+            style=l.(
+              
+              'width'='100%'
+              ,
+              'height'='60px'
+              ,
+              'lineHeight'='60px'
+              ,
+              'borderWidth'='1px'
+              ,
+              'borderStyle'='dashed'
+              ,
+              'borderRadius'='5px'
+              ,
+              'textAlign'='center'
+              ,
+              'margin'='10px'
+            )
+            ,
+            filename=''
+            
+          
+          )
+          ,
+          htmlP("", id='file_checker')
+          ,
+          htmlP(" 0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 '!' 1 ", style=l.(display='block', width='100%'))
+          
+        ))
+        
+      ))
+    
+    ))
+    
+  ))
+  
+)
+
+app$callback(
+  output=l.(id='file_checker', property='children')
+  ,
+  params=l.(input(id='upload-data', property='filename'))
+  ,
+  function(filename){
+    if(is.na(filename)|| is.null(filename)||(filename=='')){
+      return("No file ")
+    } else{
+      return(glue('Data {filename} has been successfully uploaded'))
```

Comment:
Will be honest, I don't often use `glue` 🙂 

In most cases, `sprintf` or `paste` will get you there. If you don't need special formatting that either of these can't easily provide, I'd rewrite to avoid pulling in extra packages unless it's really necessary.